### ISSUE TITLE
Segfault when misdetecting extension

### ISSUE THREAD

**Issue by barsnick** [Matthias-Wandel/jhead#14]
I can reproducibly get a segmentation fault when renaming files which Apple messed up. I have a folder full of files (mostly JPEGs) named like ```./122APPLE (403)```.

I wanted to give them back a proper extension, and a meaningful file name from the EXIF date tag:
```
jhead -n"%Y%m%d-%H%M%S_%f" './122APPLE (403)'
```
(Actually using ```find``` and ```xargs```.)

This is what happens:
```
[barsnick@goodtimes Development]$ gdb -ex r --args jhead -n"%Y%m%d-%H%M%S_%f" ./122APPLE\ \(403\) 
GNU gdb (GDB) Fedora 8.0.1-36.fc27
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from jhead...Reading symbols from /usr/lib/debug/usr/bin/jhead-3.04-1.fc27.1sunshine.x86_64.debug...done.
done.
Starting program: /usr/bin/jhead -n%Y%m%d-%H%M%S_%f ./122APPLE\ \(403\)
File './122APPLE (403)' contains no exif date stamp.  Using file date

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff782d74a in __memmove_avx_unaligned_erms () from /lib64/libc.so.6
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.26-30.fc27.x86_64
(gdb) bt
#0  0x00007ffff782d74a in __memmove_avx_unaligned_erms () from /lib64/libc.so.6
#1  0x00000000004037b8 in memmove (__len=18446744073709551614, __src=0x7fffffffe2ae, __dest=0x7fffffff9e4e) at /usr/include/bits/string_fortified.h:40
#2  DoFileRenaming (FileName=0x7fffffffe2ac "./122APPLE (403)") at jhead.c:617
#3  ProcessFile (FileName=0x7fffffffe2ac "./122APPLE (403)") at jhead.c:1247
#4  0x000000000040195c in main (argc=<optimized out>, argv=0x7fffffffdf38) at jhead.c:1756
(gdb) up 1
#1  0x00000000004037b8 in memmove (__len=18446744073709551614, __src=0x7fffffffe2ae, __dest=0x7fffffff9e4e) at /usr/include/bits/string_fortified.h:40
40        return __builtin___memmove_chk (__dest, __src, __len, __bos0 (__dest));
(gdb) l
35      }
36
37      __fortify_function void *
38      __NTH (memmove (void *__dest, const void *__src, size_t __len))
39      {
40        return __builtin___memmove_chk (__dest, __src, __len, __bos0 (__dest));
41      }
42
43      #ifdef __USE_GNU
44      __fortify_function void *
(gdb) p __len
$1 = 18446744073709551614
```

My analysis of the code told me that the ExtensionPart calculation landed at the first dot(!), while the PrefixPart calculation landed two characters later, after the ```./```. This gives a negative size for the memmove() call, which may or may not segfault. (On an old x86_32 system, it didn't segfault.)

Either the ExtensionPart calculation should begin after the PrefixPart calculation, starting at the real file name. Or if the difference is smaller than 0, just reset it to the end of the filename. I did the latter, and this is the (simple) fix I used in order to get jhead working again:

```
diff -ur jhead-3.04/jhead.c jhead-3.04-extension_segfault/jhead.c
--- jhead-3.04/jhead.c  2019-11-22 15:27:34.000000000 +0100
+++ jhead-3.04-extension_segfault/jhead.c       2020-12-22 18:00:49.392913879 +0100
@@ -584,6 +584,9 @@
 
         if (FileName[a] == '.') ExtensionPart = a;  // Remember where extension starts.
     }
+    if (ExtensionPart < PrefixPart) { // no extension found
+        ExtensionPart = strlen(FileName);
+    }
 
     if (!Exif2tm(&tm, ImageInfo.DateTime)){
         printf("File '%s' contains no exif date stamp.  Using file date\n",FileName);
```

Please see whether you can improve the extension calculation, and add some test cases. Thanks!

**Comment by Matthias-Wandel** [comment]
Thanks for figuring this one out and providing a fix.
Yes, starting the search at the last slash would be better, but then I'd 
have to add a second loop, so wouldn't be any more elegant.
  and given that you tried your fix, might as well use that one.  I just 
pushed the change onto the github repo.

Matthias

------ Original Message ------

**Comment by Matthias-Wandel** [comment]
Fixed by 11e6e871ea624427f1079df371ae3b2e6ed85dc8
